### ISSUE TITLE
Concurrent Promise is not concurrent

### ISSUE THREAD

**Issue by z0mb1ek** [reactor/reactor#512]
I have this code not working:

        Promise<Integer> p = new Promise<>(Environment.sharedDispatcher(), Environment.get()); // or just new Promise<>();
        AtomicInteger ai = new AtomicInteger();
        for (int i = 0; i < 100; i++) {
            new Thread() {
                @Override
                public void run() {
                    p.onComplete(integerPromise -> ai.incrementAndGet());
                }
            }.start();
        }
        Thread.sleep(1000);
        p.onNext(1);
        Thread.sleep(1000);
        System.out.println("Test.main " + ai);
        Thread.sleep(10000);

This is because:

        Broadcaster<Object> b = BehaviorBroadcaster.create(Environment.get());
        AtomicInteger ai = new AtomicInteger();

        for (int i = 0; i < 100; i++) {
            new Thread() {
                @Override
                public void run() {
                    b.consume(o -> ai.incrementAndGet());
                }
            }.start();
        }

        Thread.sleep(1000);
        b.onNext(1);
        Thread.sleep(1000);
        System.out.println("Test.main " + ai);
        Thread.sleep(10000);

Is it bug?

**Comment by khayredinov** [comment]
U can use ReentrantLock to fix this and same problems.
```java
lock.lock();
try {
    b.consume(o -> ai.incrementAndGet());
} finally {
    lock.unlock();
}
```

**Comment by khayredinov** [comment]
Method [```addSubscription```](https://github.com/reactor/reactor/blob/master/reactor-stream/src/main/java/reactor/rx/action/Action.java#L652) in ```Action``` not thread save, wrapping it in ```lock``` fix this.

**Comment by smaldini** [comment]
subscribe operation was not thread safe indeed, due to optimization reasons, creating a ReentrantLock for each Action is something to consider, I'll dig into it after the dispatcher/environment rationalization.

**Comment by smaldini** [comment]
To the original poster, the issue is that you reuse a same thread with the shared Dispatcher, I'm working on making it more clear tho. Will reply in details later.

**Comment by z0mb1ek** [comment]
will you reply please?

**Comment by smaldini** [comment]
Fixed and tested in reactor 2.5 integrated reactive streams commons engine
